Replace B+tree with ordered dictionary

jxsvoboda opened the pull request.

jermar commented on kernel/generic/src/mm/as.c:
```diff
@@ -1866,293 +1853,164 @@ size_t as_area_get_size(uintptr_t base)
+}
 
-success:
-	area->resident += count;
```
I think the new code does not track area->resident anymore. This probably breaks statistics.

jxsvoboda replied:
OK, I think I need to move the used page count into used_space_t.

jermar commented on kernel/generic/src/mm/backend_elf.c:
```diff
@@ -163,72 +163,65 @@ void elf_share(as_area_t *area)
 			if (!(area->flags & AS_AREA_WRITE))
 				if (base >= entry->p_vaddr &&
-				    base + P2SZ(count) <= start_anon)
+				    base + P2SZ(i + 1) <= start_anon)
```
Could this possibly overflow?

jxsvoboda replied:
I don't know but I am not changing the logic in this piece of code. Just indentation and j -> i. The diff is mismatched.

jermar commented on kernel/generic/src/mm/as.c:
```diff
@@ -2163,175 +2021,65 @@ bool used_space_insert(as_area_t *area, uintptr_t page, size_t count)
-	    "space from %p.", count, (void *) page);
-
-success:
-	area->resident -= count;
```
Again, this seems to get lost in the new version.

jxsvoboda replied:
OK.

jxsvoboda commented on kernel/generic/include/mm/as.h:
```diff
@@ -157,6 +156,27 @@ typedef struct as_pagemap {
+	size_t count;
+} used_space_ival_t;
+
+/** Map of used space in an adress space area */
```
OK

jxsvoboda commented:
I made fixes according to your comments. Number of resident pages is now tracked in used_space_t.pages. Fixed the typos and added missing documentation to functions used_space_remove_ival() and used_space_shorten_ival().

jermar commented on kernel/generic/src/mm/as.c:
```diff
@@ -1398,16 +1390,14 @@ errno_t as_area_change_flags(as_t *as, unsigned int flags, uintptr_t address)
-			used_pages += (size_t) node->value[i];
+	used_space_ival_t *ival = used_space_first(&area->used_space);
+	while (ival != NULL) {
+		used_pages += ival->count;
```
Is the computed value different from the one stored in area->used_space.pages?

jxsvoboda replied:
No, it's the same. I will change it to simply use  used_space.pages.
